spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/StreamListenerAnnotationBeanPostProcessor.java

```diff
@@ -59,56 +64,56 @@
 
-	private final DestinationResolver<MessageChannel> binderAwareChannelResolver;
+	@Autowired
+	@Lazy
+	private DestinationResolver<MessageChannel> binderAwareChannelResolver;
 
-	private final MessageHandlerMethodFactory messageHandlerMethodFactory;
+	@Autowired
+	@Lazy
+	private MessageHandlerMethodFactory messageHandlerMethodFactory;
 
-	private final Map<String, InvocableHandlerMethod> mappedBindings = new HashMap<>();
+	private final MultiValueMap<String, StreamListenerHandlerMethodMapping> mappedListenerMethods = new LinkedMultiValueMap<>();
 
 	private ConfigurableApplicationContext applicationContext;
 
 	private final List<StreamListenerParameterAdapter<?, Object>> streamListenerParameterAdapters = new ArrayList<>();
 
 	private final List<StreamListenerResultAdapter<?, ?>> streamListenerResultAdapters = new ArrayList<>();
 
-	public StreamListenerAnnotationBeanPostProcessor(DestinationResolver<MessageChannel> binderAwareChannelResolver,
-			MessageHandlerMethodFactory messageHandlerMethodFactory) {
-		Assert.notNull(binderAwareChannelResolver, "Destination resolver cannot be null");
-		Assert.notNull(messageHandlerMethodFactory, "Message handler method factory cannot be null");
-		this.binderAwareChannelResolver = binderAwareChannelResolver;
-		this.messageHandlerMethodFactory = messageHandlerMethodFactory;
-	}
+	private EvaluationContext evaluationContext;
 
 	@Override
 	@SuppressWarnings({ "rawtypes", "unchecked" })
-	public void setApplicationContext(ApplicationContext applicationContext) throws BeansException {
+	public final void setApplicationContext(ApplicationContext applicationContext) throws BeansException {
 		this.applicationContext = (ConfigurableApplicationContext) applicationContext;
-		Map<String, StreamListenerParameterAdapter> parameterAdapterMap =
-				BeanFactoryUtils.beansOfTypeIncludingAncestors(this.applicationContext, StreamListenerParameterAdapter.class);
+		Map<String, StreamListenerParameterAdapter> parameterAdapterMap = BeanFactoryUtils
+				.beansOfTypeIncludingAncestors(this.applicationContext, StreamListenerParameterAdapter.class);
 		for (StreamListenerParameterAdapter parameterAdapter : parameterAdapterMap.values()) {
 			this.streamListenerParameterAdapters.add(parameterAdapter);
 		}
-		Map<String, StreamListenerResultAdapter> resultAdapterMap =
-				BeanFactoryUtils.beansOfTypeIncludingAncestors(this.applicationContext, StreamListenerResultAdapter.class);
+		Map<String, StreamListenerResultAdapter> resultAdapterMap = BeanFactoryUtils
+				.beansOfTypeIncludingAncestors(this.applicationContext, StreamListenerResultAdapter.class);
 		this.streamListenerResultAdapters.add(new MessageChannelStreamListenerResultAdapter());
 		for (StreamListenerResultAdapter resultAdapter : resultAdapterMap.values()) {
 			this.streamListenerResultAdapters.add(resultAdapter);
 		}
+		this.evaluationContext = new StandardEvaluationContext();
```

Comment:
Can we use `IntegrationContextUtils.getEvaluationContext(BeanFactory)` here instead?